### PR TITLE
[RLlib] Fix R2D2 (torch) multi-GPU issue.

### PR DESCRIPTION
**Pull request by sven1977**
Fixes R2D2 (torch) multi-GPU issues with LSTM and attention nets.

The following yaml file would cause an error for R2D2:
```
stateless-cartpole-r2d2:
    env: ray.rllib.examples.env.stateless_cartpole.StatelessCartPole
    run: R2D2
    stop:
        episode_reward_mean: 150
        timesteps_total: 1000000
    config:
        # Works for both torch and tf.
        framework: tf
        num_workers: 0
        # R2D2 settings.
        burn_in: 20
        zero_init_states: true
        #dueling: false
        lr: 0.0005
        # Give some more time to explore.
        exploration_config:
          epsilon_timesteps: 50000
        # Wrap with an LSTM and use a very simple base-model.
        model:
            fcnet_hiddens: [64]
            fcnet_activation: linear
            #use_lstm: true
            use_lstm: true
            lstm_cell_size: 64
            max_seq_len: 20

        # Fake 2 GPUs.
        num_gpus: 2
        _fake_gpus: true
```

Same for using an attention net instead of the LSTM.

This PR fixes both issues.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/rllib_tests/learning_tests/hard_learning_tests.yaml [hunk 1]
```diff
@@ -327,13 +328,13 @@ dqn-breakoutnoframeskip-v4:
         num_gpus: 0.5
         timesteps_per_iteration: 10000
 
-es-humanoid-v2:
-    env: Humanoid-v2
+es-humanoidbulletenv-v0:
```
Comment (sven1977) [inline release/rllib_tests/learning_tests/hard_learning_tests.yaml:331]: Wrong env specified (must be PyBullet, not MuJoCo).

File: release/rllib_tests/learning_tests/hard_learning_tests.yaml [hunk 2]
```diff
@@ -327,13 +328,13 @@ dqn-breakoutnoframeskip-v4:
...
+    env: HumanoidBulletEnv-v0
     run: ES
     # Minimum reward and total ts (in given time_total_s) to pass this test.
     pass_criteria:
-        episode_reward_mean: 6000.0
-        timesteps_total: 10000000
+        episode_reward_mean: 100.0
```
Comment (sven1977) [inline release/rllib_tests/learning_tests/hard_learning_tests.yaml:336]: This should work/learn now.

File: rllib/BUILD [hunk 3]
```diff
@@ -663,6 +663,26 @@ py_test(
     args = ["--yaml-dir=tuned_examples/dqn", "--framework=torch"]
 )
 
+py_test(
```
Comment (sven1977) [inline rllib/BUILD:666]: R2D2 added to fake-GPU (per-PR) tests.

File: rllib/agents/dqn/r2d2_tf_policy.py [hunk 4]
```diff
@@ -45,7 +45,8 @@ def build_r2d2_model(policy: Policy, obs_space: gym.spaces.Space,
     model = build_q_model(policy, obs_space, action_space, config)
 
     # Assert correct model type.
-    assert model.get_initial_state() != [], \
+    assert (model.get_initial_state() != [] or
```
Comment (sven1977) [inline rllib/agents/dqn/r2d2_tf_policy.py:48]: Attention nets don't publish init state. They use the traj. view API only.

File: release/rllib_tests/learning_tests/hard_learning_tests.yaml (hunk 1 above)
Comment (gjoliver) [reply release/rllib_tests/learning_tests/hard_learning_tests.yaml:331]: sounds good

File: release/rllib_tests/learning_tests/hard_learning_tests.yaml (hunk 2 above)
Comment (gjoliver) [reply release/rllib_tests/learning_tests/hard_learning_tests.yaml:336]: ok

File: rllib/BUILD (hunk 3 above)
Comment (gjoliver) [reply rllib/BUILD:666]: nice

File: rllib/agents/dqn/r2d2_tf_policy.py (hunk 4 above)
Comment (gjoliver) [reply rllib/agents/dqn/r2d2_tf_policy.py:48]: Can we add this comment as a comment in the code? Will that make the intention of the 2nd condition clearer here?
Thanks.

File: rllib/agents/dqn/r2d2_torch_policy.py [hunk 5]
```diff
@@ -53,7 +53,8 @@ def build_r2d2_model_and_distribution(
         policy, obs_space, action_space, config)
 
     # Assert correct model type.
-    assert model.get_initial_state() != [], \
+    assert (model.get_initial_state() != [] or
```
Comment (gjoliver) [inline rllib/agents/dqn/r2d2_torch_policy.py:56]: same here, I guess.

File: rllib/policy/sample_batch.py [hunk 6]
```diff
@@ -832,6 +832,8 @@ def _slice(self, slice_: slice):
         """
         start = slice_.start or 0
         stop = slice_.stop or len(self)
```
Comment (gjoliver) [inline rllib/policy/sample_batch.py:834]: seems like this line plus the if block you just added is basically:

stop = math.min(slice_.stop, len(self))

and then maybe we can add a quick comment about capping it at len(self)?

File: rllib/tuned_examples/dqn/stateless-cartpole-r2d2-fake-gpus.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
+stateless-cartpole-r2d2:
+    env: ray.rllib.examples.env.stateless_cartpole.StatelessCartPole
+    run: R2D2
+    stop:
+        episode_reward_mean: 150
+        timesteps_total: 1000000
+    config:
+        # Works for both torch and tf.
+        framework: tf
+        num_workers: 0
```
Comment (gjoliver) [inline rllib/tuned_examples/dqn/stateless-cartpole-r2d2-fake-gpus.yaml:10]: just to double check, should we add a non-local worker here?

File: rllib/policy/sample_batch.py [hunk 8]
```diff
@@ -843,7 +845,8 @@ def _slice(self, slice_: slice):
                     for _ in range(l):
                         self._slice_map.append((i, sum_))
                     sum_ += l
-                self._slice_map.append((len(self[SampleBatch.SEQ_LENS]), sum_))
+                self._slice_map.append((len(self[SampleBatch.SEQ_LENS]) - 1,
```
Comment (gjoliver) [inline rllib/policy/sample_batch.py:848]: actually, I am not quite sure if I get this change here.
why are we leaving out the last seq?
can you help comment a bit?
thanks.

**Review by gjoliver — COMMENTED** [review]
really nice fix. have a few minor questions I am hoping you can help educate me a little bit :)
thanks!

File: rllib/tuned_examples/dqn/stateless-cartpole-r2d2-fake-gpus.yaml (hunk 7 above)
Comment (sven1977) [reply rllib/tuned_examples/dqn/stateless-cartpole-r2d2-fake-gpus.yaml:10]: You are probably right. I think I tried to reduce the number of CPUs this needs on the (expensive) GPU machine, but this is fake-GPUs, so it's fine. Will change.

File: rllib/policy/sample_batch.py (hunk 8 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:848]: Will do.

File: rllib/agents/dqn/r2d2_tf_policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/agents/dqn/r2d2_tf_policy.py:48]: done

File: rllib/agents/dqn/r2d2_torch_policy.py (hunk 5 above)
Comment (sven1977) [reply rllib/agents/dqn/r2d2_torch_policy.py:56]: done

File: rllib/policy/sample_batch.py (hunk 6 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:834]: done

**Review by gjoliver — APPROVED** [review]
I get it now. as always, thanks for the replies :)

**Pull request merged**